Fix part of #1366: Services for creator dashboard

526avijitgupta opened the pull request:
Tasks to be done as a part of this PR:

- [x] Get and update per-exploration stats. (UPDATE: Can be directly done obtained in the controller using existing methods - `stats_services.get_exploration_stats(exp_id, exp_version)` for num_plays and ratings are already being sent in the serializer)
- [x] Method to get values of user-stats model
- [x] Unit tests for this method

526avijitgupta commented:
@wxyxinyu @seanlip Can you please take a first review pass. I guess writing more unit tests for the service which fetches the user stats model should complete the tasks of this PR as well ? (Take a look at the comment above)

wxyxinyu commented on core/domain/user_services_test.py:
```diff
@@ -18,15 +18,38 @@
 
 from google.appengine.api import urlfetch
 
+class ModifiedUserStatsAggregator(user_jobs_continuous.UserStatsAggregator):
```
Isn't this the same as the one used in user_jobs_continuous_test? Couldn't you just import user_jobs_continuous_test directly instead of copying the code here?

seanlip replied:
+1

526avijitgupta replied:
Done

wxyxinyu commented:
Looks good generally. Could you figure out why the tests are failing?

seanlip commented on core/domain/user_services_test.py:
```diff
@@ -472,6 +495,38 @@ def test_contribution_msec_does_not_change_if_collection_unpublished(self):
+            feconf.PLAY_TYPE_NORMAL)
+        ModifiedUserStatsAggregator.start_computation()
+        self.process_and_flush_pending_tasks()
+        self.assertEquals(
```
You should do a "before" check as well by calling the same function before the computation is run.

526avijitgupta replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -169,6 +169,17 @@ def get_versions_for_exploration_stats(exploration_id):
+    exploration = exp_services.get_exploration_by_id(exploration_id)
+    state_names = exploration.states.keys()
+
+    return sum(stats_domain.StateRuleAnswerLog.get(
```
Is this slow? If you're going to end up doing it for multiple explorations at a time, you may need to batch things. Otherwise you'll be doing N get() calls in sequence rather than 1 get_multi(); the former is much slower.

/cc @BenHenning as this intersects with his answer-migration work

526avijitgupta replied:
I agree it can be heavy on the system. Also, I think that the current `get_multi` method inside `StateRuleAnswerLog` class (in core.domain.stats_domain) will not will not help in this case. Can you please elaborate what did you meant by batching things. ?

wxyxinyu replied:
The `get_multi` method will help, at it means you will only be doing 1 call to the database, rather than N calls (where N is the number of states). Each `get` call accesses the database once.

526avijitgupta replied:
Used `get_multi`. see latest commit

seanlip commented:
Took a review pass. I observe that this PR is actually made up of two independent components, so if there end up being issues with the get_unresolved_answers_count part, we could break this into two PRs and submit the other first.

526avijitgupta commented:
@wxyxinyu Fixed failing tests and addressed your review comment as well.

@seanlip Addressed all but one review comments (left a comment about the one which is left). I agree about breaking this up into 2 PRs, since get_total_unresolved_answers_for_exploration part looks harder that I expected it to be!

526avijitgupta commented:
@seanlip @wxyxinyu Reverted the commits which referred to unresolved answers part of the dashboard. Opened https://github.com/oppia/oppia/pull/2101 for that.
This should be ready for merge, I guess ?

wxyxinyu commented:
lgtm
